File: go-backend-interview-prep/10_design_patterns_in_go/06_decorator_middleware/main.py
"""Decorator / middleware pattern -- Python equivalent of the Go example.

Each decorator wraps a sender, adds behavior, delegates to the original.
"""

import time


# --- Base implementation ---

class EmailSender:
    def __init__(self):
        self.fail_count = 0

    def send(self, to: str, message: str) -> None:
        self.fail_count += 1
        if self.fail_count <= 2:
            raise RuntimeError("temporary failure")
        print(f"  [EMAIL] To: {to} | {message}")


class SimpleSender:
    def send(self, to: str, message: str) -> None:
        print(f"  [SEND] {to}: {message}")


# --- Decorator: logging ---

class LoggingSender:
    def __init__(self, next_sender):
        self.next = next_sender

    def send(self, to: str, message: str) -> None:
        print(f"  [LOG] sending to {to} at {time.strftime('%H:%M:%S')}")
        try:
            self.next.send(to, message)
            print("  [LOG] success")
        except Exception as e:
            print(f"  [LOG] error: {e}")
            raise


# --- Decorator: retry ---

class RetrySender:
    def __init__(self, next_sender, max_retries: int = 3):
        self.next = next_sender
        self.max_retries = max_retries

    def send(self, to: str, message: str) -> None:
        last_err = None
        for attempt in range(1, self.max_retries + 1):
            try:
                self.next.send(to, message)
                return
            except Exception as e:
                last_err = e
                print(f"  [RETRY] attempt {attempt}/{self.max_retries} failed: {e}")
        raise RuntimeError(f"all {self.max_retries} retries failed: {last_err}")


# --- Decorator: rate limiter ---

class RateLimitSender:
    def __init__(self, next_sender, delay: float):
        self.next = next_sender
        self.delay = delay
        self.last_sent = 0.0

    def send(self, to: str, message: str) -> None:
        now = time.monotonic()
        elapsed = now - self.last_sent
        if self.last_sent > 0 and elapsed < self.delay:
            wait = self.delay - elapsed
            print(f"  [RATE] throttling {wait*1000:.0f}ms...")
            time.sleep(wait)
        self.last_sent = time.monotonic()
        self.next.send(to, message)


def main() -> None:
    print("=== Decorator / Middleware Pattern ===")
    print()

    # Layer: logging -> retry -> email (fails first 2 times)
    print("--- Logging + Retry (email fails first 2 times) ---")
    base = EmailSender()
    sender = LoggingSender(RetrySender(base, max_retries=3))
    try:
        sender.send("alice@example.com", "Order shipped!")
    except RuntimeError as e:
        print(f"  Final error: {e}")
    print()

    # Rate limited
    print("--- Rate limited sender ---")
    limited = RateLimitSender(LoggingSender(SimpleSender()), delay=0.2)
    for i in range(3):
        limited.send("bob", f"msg-{i}")
    print()

    print("Key: each decorator has .send(), wraps another .send().")
    print("     Compose: LoggingSender(RetrySender(RateLimitSender(base))).")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/10_design_patterns_in_go/06_decorator_middleware/main.go
package main

import (
	"errors"
	"fmt"
	"time"
)

// --- Core interface ---

type Sender interface {
	Send(to, message string) error
}

// --- Base implementation ---

type EmailSender struct {
	failCount int // simulate failures for retry demo
}

func (e *EmailSender) Send(to, message string) error {
	e.failCount++
	if e.failCount <= 2 {
		return errors.New("temporary failure")
	}
	fmt.Printf("  [EMAIL] To: %s | %s\n", to, message)
	return nil
}

// --- Decorator: logging ---

type loggingSender struct {
	next Sender
}

func WithLogging(next Sender) Sender {
	return &loggingSender{next: next}
}

func (l *loggingSender) Send(to, message string) error {
	fmt.Printf("  [LOG] sending to %s at %s\n", to, time.Now().Format("15:04:05"))
	err := l.next.Send(to, message)
	if err != nil {
		fmt.Printf("  [LOG] error: %v\n", err)
	} else {
		fmt.Printf("  [LOG] success\n")
	}
	return err
}

// --- Decorator: retry ---

type retrySender struct {
	next       Sender
	maxRetries int
}

func WithRetry(next Sender, maxRetries int) Sender {
	return &retrySender{next: next, maxRetries: maxRetries}
}

func (r *retrySender) Send(to, message string) error {
	var err error
	for attempt := 1; attempt <= r.maxRetries; attempt++ {
		err = r.next.Send(to, message)
		if err == nil {
			return nil
		}
		fmt.Printf("  [RETRY] attempt %d/%d failed: %v\n", attempt, r.maxRetries, err)
	}
	return fmt.Errorf("all %d retries failed: %w", r.maxRetries, err)
}

// --- Decorator: rate limiter (simple) ---

type rateLimitSender struct {
	next    Sender
	delay   time.Duration
	lastSent time.Time
}

func WithRateLimit(next Sender, delay time.Duration) Sender {
	return &rateLimitSender{next: next, delay: delay}
}

func (r *rateLimitSender) Send(to, message string) error {
	since := time.Since(r.lastSent)
	if since < r.delay {
		wait := r.delay - since
		fmt.Printf("  [RATE] throttling %s...\n", wait.Round(time.Millisecond))
		time.Sleep(wait)
	}
	r.lastSent = time.Now()
	return r.next.Send(to, message)
}

func main() {
	fmt.Println("=== Decorator / Middleware Pattern ===")
	fmt.Println()

	// Layer decorators: logging -> retry -> email sender
	fmt.Println("--- Logging + Retry (email fails first 2 times) ---")
	base := &EmailSender{}
	sender := WithLogging(WithRetry(base, 3))
	err := sender.Send("alice@example.com", "Order shipped!")
	if err != nil {
		fmt.Printf("  Final error: %v\n", err)
	}
	fmt.Println()

	// Demonstrate rate limiting
	fmt.Println("--- Rate limited sender ---")
	simpleSender := WithRateLimit(
		WithLogging(&simplePrintSender{}),
		200*time.Millisecond,
	)
	for i := 0; i < 3; i++ {
		simpleSender.Send("bob", fmt.Sprintf("msg-%d", i))
	}
	fmt.Println()

	fmt.Println("Key: each decorator implements Sender, wraps another Sender.")
	fmt.Println("     Compose: WithLogging(WithRetry(WithRateLimit(base))).")
	fmt.Println("     Same pattern as Go HTTP middleware.")
}

type simplePrintSender struct{}

func (s *simplePrintSender) Send(to, message string) error {
	fmt.Printf("  [SEND] %s: %s\n", to, message)
	return nil
}
